Arrival rate executors review (#19)
grafana/k6-learn: Modules/III-k6-Intermediate/08-Setting-load-profiles-with-executors/Constant-Arrival-Rate-Exercises.md

As noted in [Setting load profiles with executors](../08-Setting-load-profiles-with-executors.md#Constant-Arrival-Rate), the _Constant Arrival Rate_ executor has a primary focus on the _iteration rate_ being applied over a specified timeframe.

This is a scenario typically seen with [load testing an API](https://k6.io/docs/testing-guides/api-load-testing/) when there is a need to simulate a constant request rate for a particular API endpoint.

## Exercises

> For our exercises, we're going to start by using a basic script that simply performs an HTTP request on each test iteration and aims at achieving 50 requests per time unit (which defaults to 1 second). We're providing some console output as things change.

Reviewer: ```suggestion
Our exercises start with a basic script. It runs an HTTP request on each test iteration and tries to achieve 50 requests per time unit (which defaults to 1 second). We provide some console output as things change.
```

Reviewer: Shortening sentences. Also, avoid "simple" as an adjective because it's subjective.

### Creating our script

Let's begin by implementing our test script with the minimally required configuration. Create a file named _test.js_ with the following content:

```js
import http from 'k6/http';

export const options = {
  scenarios: {
    k6_workshop: {
      executor: 'constant-arrival-rate',
      rate: 50,
      duration: '30s',
      preAllocatedVUs: 2,
    },
  },
};

export default function () {
  console.log(`[VU: ${__VU}, iteration: ${__ITER}] Starting iteration...`);
  http.get('https://test.k6.io/contacts.php');
}
```

We're starting with the bare minimum to use the executor. Compared to previous executors, this has a bit more required configuration beyond the usual `executor` itself.

> Reviewing the [options in the documentation](https://k6.io/docs/using-k6/scenarios/executors/constant-arrival-rate/), we see that the `rate` and `duration` are now required. This makes sense given the focus of this executor is achieving and maintaining the specified _iteration rate_ over the provided timeframe. 

Reviewer: ```suggestion
As noted in the [`constant-arrival-rate` documentation](https://k6.io/docs/using-k6/scenarios/executors/constant-arrival-rate/),  `rate` and `duration` are now required. This makes sense given the focus of this executor is to acheive and maintain the specified _iteration rate_ over the provided timeframe. 
```

     vus_max........................: 2      min=2       max=2
```

Our test ran successfully. However, a closer inspection of the results shows that our actual results are not as intended.

Looking at the output, we see the following:

```bash
WARN[0000] Insufficient VUs, reached 2 active VUs and cannot initialize more  executor=constant-arrival-rate scenario=k6_workshop
```

What happened here? 

Remember the `preAllocatedVUs` setting we glossed over earlier? With this setting, we simply told k6 to start the test with 2 virtual users; after a few iterations, k6 was able to determine that it would **not** be able to achieve our desired iteration rate (50 iterations/s).

This fact is evident by looking at the `iterations` value in the test summary; our test was only able to attain a rate of 13.61 iterations per second and we wanted 50.

### Adjusting preallocated virtual users

Restating once again, the _Constant Arrival Rate_ executor is focused on the _iteration rate_ over a time frame. k6 aims to eliminate the need to be overly concerned about the actual number of users required to achieve such a rate. However, we still need to give our script enough VUs to achieve that rate. 

> From our example above, we have that our request duration or latency is, on average, 132.35ms, and the 95 percentile is around 165.43ms. With just 2 VUs (`preAllocatedVUs`), in a very optimistic scenario, we cannot expect more than `2 VUs / 0.132 s = 15.15 iterations/s`.

Reviewer: Try to avoid "above" and "below" and other directional language. This is for inclusivity.

     http_req_waiting...............: avg=136.81ms min=101.73ms med=114.5ms  max=862.74ms p(90)=146.57ms p(95)=184.31ms
     http_reqs......................: 1501   49.213783/s
     iteration_duration.............: avg=141.27ms min=101.88ms med=114.83ms max=862.9ms  p(90)=150.38ms p(95)=407.84ms
     iterations.....................: 1501   49.213783/s
     vus............................: 25     min=25      max=25
     vus_max........................: 25     min=25      max=25
```

Reviewing the output now, we see that the desired rate is more closely achieved at 49.21 iterations per second (for the overall test).

You might be tempted to set a lower value for [`preAllocatedVUs` and use `maxVUs` option](https://k6.io/docs/using-k6/scenarios/executors/constant-arrival-rate/) to autoscale the test. `maxVUs` is the maximum number of VUs to allow during the test run, which defaults to `preAllocatedVUs` when not set. However, it's usually best to leave the default value for `maxVUs` and increase the `preAllocatedVUs`. In this way, the VUs are allocated before the test starts and will be used when (and only if) necessary.

> Allocating VUs in the middle of the test can be costly in terms of CPU resources in the load generator instance, and can skew the tests. Preallocating VUs will allow the test to start with all the VUs available with no need to wait to allocate more when needed in the middle of the test run.

Reviewer: Also I try to avoid "allows" since it some readers find it paternalizing. It's not really about permission as much as it is about a clear state of affairs.
